Allow CustomTarget's to be indexed

dcbaker opened the pull request:
This allows a CustomTarget to be indexed, and the resulting indexed
value (a CustomTargetIndex type), to be used as a source in other
targets. This will confer a dependency on the original target, but only
inserts the source file returning by index the original target's
outputs. This can allow a CustomTarget that creates both a header and a
code file to have it's outputs split, for example.

The larger detail is that it implements the concept of a protocol, which
is borrowed from python terminology for implementing things like
indexing, adding, etc. It adds the index_protocol which the interpreter
will evaluate for any object that has that method.

Fixes #1470

nirbheek commented on mesonbuild/build.py:
```diff
@@ -1735,6 +1738,25 @@ def validate_cross_install(self, environment):
+
+    def __init__(self, target, output):
+        self.target = target
+        self.output = output
```
This also needs a `__repr__` implementation because we sometimes print internal meson objects in exceptions and such, and it's good to have a debuggable representation then.

dcbaker replied:
can do.

nirbheek commented on mesonbuild/build.py:
```diff
@@ -1676,6 +1676,9 @@ def get_generated_sources(self):
         return "@cus"
 
+    def get_index_protocol(self, index):
+        return CustomTargetIndex(self, self.outputs[index])
```
Another way to do this would be to implement `__getitem__` and implement `__setitem__` and `__delitem__` to raise `AssertionError`s. That would mean that these objects would also be iterable from Python, but perhaps that's a good thing.

dcbaker replied:
I wasn't sure. I thought about just implementing `__getitem__` but I feel like that's a bit of a layering violation (I think that meson's lack of a ListHolder is also a bit of a layering violation)

nirbheek commented on mesonbuild/interpreter.py:
```diff
@@ -582,6 +587,9 @@ def __repr__(self):
         return self.interpreter.backend.get_target_filename_abs(self.held_object)
 
+    def index_protocol(self, args, kwargs):
+        return CustomTargetIndexHolder(self.held_object.get_index_protocol(args[0]))
```
Same here, can implement `__getitem__`, etc, and return `InterpreterException` for set/del.

nirbheek replied:
We also need a `FailureTests` unit test for setting by index.

dcbaker replied:
But that would mean implementing `__getitem__` with a non-standard signature. That seems *really* mean.

nirbheek commented on mesonbuild/interpreterbase.py:
```diff
@@ -369,14 +369,19 @@ def evaluate_plusassign(self, node):
-            raise InterpreterException('Tried to index a non-array object.')
+        if not (isinstance(iobject, list) or hasattr(iobject, 'index_protocol')):
+            raise InterpreterException(
+                'Tried to index an object that doesn\'t support indexing.')
```
Then this can just be `hasattr(iobject, '__getitem__')`, which will work for all indexable objects.

nirbheek commented on mesonbuild/interpreterbase.py:
```diff
@@ -369,14 +369,19 @@ def evaluate_plusassign(self, node):
+            if isinstance(iobject, list):
+                return iobject[index]
+            else:
+                return iobject.index_protocol([index], {})
```
... and this special case will go away.

dcbaker replied:
On the other hand, I really don't think that allowing python's list implementation to be directly exposed in meson is good, it's a layering violation. Wouldn't it be better to have a meson `List` and `ListHolder` object so that we don't rely on python implementation details?

nirbheek replied:
Yes, it is a layering violation, but we do such things in numerous places because there are advantages to that, namely:

1. Our corner-case behaviour is sane and 'non-surprising'/'expected' by default without too much work by us
2. The amount of work needed to implement features is a lot less
3. Parser behavior is well-documented because it is often the same as Python
4. Parsing is faster because directly using python builtins is faster than pure-python code
5. Porting Meson to another language is actually easier, because of (3)

We do have to be careful to not expose strange edge-cases that are Python-specific which may tie us to the language, but I don't think this is one of them.

nirbheek commented on test cases/common/161 index customtarget/meson.build:
```diff
@@ -0,0 +1,43 @@
+foreach n : ['1', '2', '3']
+  lib = static_library(
+    'lib@0@'.format(n),
+    ['lib.c', gen[1]],
```
Wouldn't it be nicer if this only took the header and caused a build error if `gen[1]` returned both the header and the code?

dcbaker replied:
that would certainly be easier.

dcbaker commented:
I didn't look at the modules, I'll look at those too.

dcbaker commented:
Okay, I think I've addressed the gnome module (none of the other modules check for CustomTarget). I'v also added the failing test, and I added a test (and fixed) using a generated source from another directory (The first version missed a check that was necissary to get -I)

dcbaker commented:
Thanks!
